### ISSUE TITLE
Repair clusters by reducing cluster size

### ISSUE THREAD

**Issue by lni** [lni/dragonboat#178]
Current version of dragonboat provides a tool called ImportSnapshot to repair clusters that permanently lost their quorums. For example, a 3 nodes cluster becomes unavailable when it loses 2 nodes permanently. The ImportSnapshot tool repairs such cluster by manipulating the membership records to include new nodes into the cluster, a snapshot previously exported from the cluster is imported to those new nodes so they can start from a recent point in time. 

This existing approach tries to prevent further degradation by forcing new nodes to be included. The drawbacks are - 

1. an exported snapshot is required, new nodes must be available   
2. importing a snapshot can be time consuming

An alternative approach is to provide a tool to rewrite the membership records on survived node to only include those survived nodes. In the above example, the only remaining node would become the only member of the cluster.

**Comment by kolinfluence** [comment]
is this part of dragonboat 4.0?

possible to provide the milestone / roadmap for 4.0?

**Comment by lni** [comment]
No, there is no plan yet to implement this. The above mentioned tool currently available works fine. 

Also note that this is not for normal operation, it is only used when you've lost your quorum (majority of nodes). If you have the feeling that you'd be put into the situation to deal with such issues, you should probably increase your replica count (e.g. from 3 to 5).
